### PR TITLE
Add devcontainer to CI pipelines

### LINKED ISSUE #504: Add devcontainer to CI pipelines
We can now run CI pipelines in devcontainers, just like we do with linux/windows/mac.

Here's more info:
https://github.com/devcontainers/ci

Let's add this to our templates tests.

### PR DESCRIPTION
**Pull request by hemarina**
Fix #504

### CODE REVIEW

File: eng/pipelines/templates/steps/template-test-run-job.yml [hunk 1]
```diff
@@ -43,27 +48,135 @@ steps:
...
-    ARM_CLIENT_ID: $(arm-client-id)
-    ARM_CLIENT_SECRET: $(arm-client-secret)
-    ARM_TENANT_ID: $(arm-tenant-id)
+- pwsh: |
+    git config --global user.name "azure-sdk"
+    git config --global user.email "azuresdk@microsoft.com"
+    echo "https://azure-sdk:$(azuresdk-github-pat)@github.com" > ~/.git-credentials
+    git config --global credential.helper "store"
+    Write-Host "##vso[task.setvariable variable=folderPath;]$pwd"
+    git clone --branch pr/1000 https://github.com/$(TemplateName).git/ temp
```
Comment (vhvb1989) [inline eng/pipelines/templates/steps/template-test-run-job.yml:57]: why is it `pr/1000` ?

File: eng/pipelines/templates/steps/template-test-run-job.yml [hunk 2]
```diff
@@ -43,27 +48,135 @@ steps:
...
-  displayName: Test templates
-  env:             
-    # Required secrets for Terraform service principal authentication
-    ARM_CLIENT_ID: $(arm-client-id)
-    ARM_CLIENT_SECRET: $(arm-client-secret)
-    ARM_TENANT_ID: $(arm-tenant-id)
+- pwsh: |
+    git config --global user.name "azure-sdk"
+    git config --global user.email "azuresdk@microsoft.com"
+    echo "https://azure-sdk:$(azuresdk-github-pat)@github.com" > ~/.git-credentials
```
Comment (vhvb1989) [inline eng/pipelines/templates/steps/template-test-run-job.yml:54]: do we need to use `env:` to map the `azuresdk-github-pat` ?  @danieljurek

File: eng/pipelines/templates/steps/template-test-run-job.yml [hunk 3]
```diff
@@ -43,27 +48,135 @@ steps:
...
+      azd version
+      azd login \
+        --client-id "$(arm-client-id)" \
+        --client-secret "$(arm-client-secret)" \
+        --tenant-id "$(arm-tenant-id)"
+
+      sudo add-apt-repository ppa:gstreamer-developers/ppa
+      sudo apt-get update
+      sudo apt-get install gstreamer1.0*
+      sudo apt-get install -y gstreamer1.0-libav libnss3-tools libatk-bridge2.0-0 libcups2-dev libxkbcommon-x11-0 libxcomposite-dev libxrandr2 libgbm-dev libgtk-3-0
```
Comment (vhvb1989) [inline eng/pipelines/templates/steps/template-test-run-job.yml:81]: might want to add a comment here about why do we need all this libs

File: eng/pipelines/templates/steps/template-test-run-job.yml [hunk 4]
```diff
@@ -43,27 +48,137 @@ steps:
...
+              for TEMPLATE in "${!ENV_TEMPLATE_MAP[@]}"; do
+                  (cleanupTemplate "$TEMPLATE" "$BRANCH_NAME" "${ENV_TEMPLATE_MAP[$TEMPLATE]}" || continue) &
+              done
+
+              wait
+          fi
+
+          echo ""
+          echo "Done!"
+      else
```
Comment (weikanglim) [inline eng/pipelines/templates/steps/template-test-run-job.yml:168]: We can remove this entire part since the template name is always specified.

File: templates/common/.devcontainer/devcontainer.json/csharp/devcontainer.json [hunk 5]
```diff
@@ -19,6 +19,9 @@
         "ghcr.io/devcontainers/features/node:1": {
             "version": "16",
             "nodeGypDependencies": false
+        },
+        "ghcr.io/rchaganti/vsc-devcontainer-features/azurebicep:1":{
+            "version":"latest"
         }
```
Comment (weikanglim) [inline templates/common/.devcontainer/devcontainer.json/csharp/devcontainer.json:25]: We recently merged merged #1210. This means even with bicep installed, `azd` will still acquire it's own copy of `bicep`. Thus, we can simply revert this change.

**Review by weikanglim — CHANGES_REQUESTED** [review]
Everything looks good. Just waiting for the removal of the bicep devcontainer feature given recent changes.

**Review by weikanglim — APPROVED** [review]
LGTM

File: eng/pipelines/templates/steps/template-test-run-job.yml (hunk 1 above)
Comment (weikanglim) [inline eng/pipelines/templates/steps/template-test-run-job.yml]: This need so to be changed to $(TemplateBranchName)?

File: eng/pipelines/templates/steps/template-test-run-job.yml (hunk 2 above)
Comment (danieljurek) [reply eng/pipelines/templates/steps/template-test-run-job.yml:54]: We don't need `env` in this case. If the secret is associated in the pipeline the script will be generated with the secret written in... 

@hemarina -- Use this template instead of duplicating the git credential setting code: https://github.com/Azure/azure-dev/blob/main/eng/pipelines/templates/steps/set-git-credentials.yml

You can include it with: 

```yaml
- template: /eng/pipelines/templates/steps/set-git-credentials.yml
```

You'll need to break the `git checkout` into a separate task.

File: eng/pipelines/templates/steps/template-test-run-job.yml [hunk 6]
```diff
@@ -43,27 +48,100 @@ steps:
...
-    # Required secrets for Terraform service principal authentication
-    ARM_CLIENT_ID: $(arm-client-id)
-    ARM_CLIENT_SECRET: $(arm-client-secret)
-    ARM_TENANT_ID: $(arm-tenant-id)
+- pwsh: |
+    git config --global user.name "azure-sdk"
+    git config --global user.email "azuresdk@microsoft.com"
+    echo "https://azure-sdk:$(azuresdk-github-pat)@github.com" > ~/.git-credentials
+    git config --global credential.helper "store"
+    Write-Host "##vso[task.setvariable variable=folderPath;]$pwd"
```
Comment (danieljurek) [inline eng/pipelines/templates/steps/template-test-run-job.yml:56]: We don't need to use `folderPath` here. The default location of all commands is `$(Build.SourcesDirectory)` https://learn.microsoft.com/en-us/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml

Naturally there are some caveats, but in this case since the pipeline is only checking out a [single repo](https://github.com/hemarina/azure-dev/blob/fix-504/eng/pipelines/templates/jobs/run-template-tests.yml#L68-L69) so the default working directory is the root of the `azure-dev` repo as it is checked out on the machine.

**Review by vhvb1989 — APPROVED** [review]
